rpcc/conn.go

```diff
@@ -1,536 +1,571 @@
+			ws.NetDial = func(network, addr string) (net.Conn, error) {
+				conn, err := (&net.Dialer{}).DialContext(ctx, network, addr)
+				// Use writeLimiter to avoid writing fragmented
+				// websocket messages. We're not accounting for
+				// the header length here because it varies, as
+				// a result we might block some valid writes
+				// that are a few bytes too large.
+				return &writeLimiter{
+					timeout: c.dialOpts.timeout,
+					limit:   ws.WriteBufferSize,
+					Conn:    conn,
+				}, err
+			}
+
+			wsConn, _, err := ws.Dial(addr, nil)
+			if err != nil {
+				return nil, err
+			}
+
+			if ws.EnableCompression {
+				c.compressionLevel = wsConn.SetCompressionLevel
+			}
+
+			return &wsReadWriteCloser{wsConn: wsConn}, nil
+		}
+	}
+
+	c.conn, err = netDial(ctx, target)
+	if err != nil {
+		return nil, err
+	}
+	newCodec := c.dialOpts.codec
+	if newCodec == nil {
+		newCodec = func(conn io.ReadWriter) Codec {
+			return &jsonCodec{
+				enc: json.NewEncoder(conn),
+				dec: json.NewDecoder(conn),
+			}
+		}
+	}
+	c.codec = newCodec(c.conn)
+
+	recvDone := func(err error) {
+		// When we receive Inspector.detached the remote will close
+		// the connection afterwards and recvDone will return. Maybe
+		// we could give the user time to react to the event before
+		// closing?
+		// TODO(mafredri): Do we want to close here, like this?
+		c.close(err)
+	}
+	go c.recv(c.notify, recvDone)
+
+	return c, nil
+}
+
+// writeLimiter wraps a net.Conn and prevents writes of greater length
+// than limit. Works around Chrome's lack of support for large or
+// fragmented websocket messages and prevents sudden termination of the
+// websocket connection. Gives the user an actionable error message when
+// writes exceed limit.
+type writeLimiter struct {
+	timeout time.Duration
+	limit   int
+	net.Conn
+}
+
+// BUG(mafredri): Chrome does not support websocket fragmentation
+// (continuation messages) or messages that exceed 1MB in size.
+// This limit was bumped in more recent versions of Chrome which can
+// receive messages up to 100MB in size.
+// See https://github.com/mafredri/cdp/issues/4 and
+// https://github.com/ChromeDevTools/devtools-protocol/issues/24.
+func (c *writeLimiter) Write(b []byte) (n int, err error) {
+	if len(b) > c.limit {
+		return 0, errors.New("rpcc: message too large (increase write buffer size or enable compression)")
+	}
+	if err = c.setWriteDeadline(); err != nil {
+		return 0, err
+	}
+	return c.Conn.Write(b)
+}
+
+func (c *writeLimiter) Read(b []byte) (n int, err error) {
+	if err = c.setReadDeadline(); err != nil {
+		return 0, err
+	}
+	return c.Conn.Read(b)
+}
+
+func (c *writeLimiter) setWriteDeadline() error {
+	if c.timeout < 1 {
+		return nil
+	}
+	return c.Conn.SetWriteDeadline(time.Now().Add(c.timeout)) // add by biter
+}
+
+func (c *writeLimiter) setReadDeadline() error {
+	if c.timeout < 1 {
+		return nil
+	}
+	return c.Conn.SetReadDeadline(time.Now().Add(c.timeout)) // add by biter

```

Comment:
These comments could be removed, it will be visible in the git history :).

Reply:
Of course!
This is for me, to distinguish your code from mine :)